### PR TITLE
Add JWT claim decryption for audit

### PR DESCRIPTION
**Pull request by NickGrantX**
# Motivation and Context
We need to know which users are doing what, via the ROps UI, so that we have an audit trail.

# What has changed
Added JWT claim decryption gubbins.

# How to test?
Once you have set up your environment with all the IAP stuff so that you can get to the ROps UI, you need to go to _Security -> Identity-Aware Proxy_ and click on _Signed Header JWT_ and copy the value.

Set up the secret: `kubectl create secret generic iap-audiences --from-literal=rops-ui=<VALUE>`

Then majick happens - go to the ops UI and marvel at the wonder.

# Links
Trello: https://trello.com/c/TlFn3mv1

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -1,16 +1,42 @@
...
+def get_iap_public_key(key_id):
+    if key_id not in get_iap_public_key.cache:
+        resp = requests.get('https://www.gstatic.com/iap/verify/public_key')
+        resp.raise_for_status()
+        get_iap_public_key.cache = resp.json()
+
+    return get_iap_public_key.cache[key_id]
+
+
+get_iap_public_key.cache = {}
```
Comment (AdamHawtin) [inline run.py:33]: `functools.lru_cache` is a nice built in way of caching function results like this

**Review by AdamHawtin — CHANGES_REQUESTED** [review]
Instructions for setting up the IAP audience config need to be documented somewhere.

**Review by AdamHawtin — CHANGES_REQUESTED** [review]
The testing instructions say to insert the IAP_AUDIENCE into the config for ops-ui, I assume you mean editing the deployed YAML? That works for testing it now but in CI for example it'll get overriden every time it's redeployed, it needs to be stored/fetched from a configmap. That then gives us the issue that it won't start up without it, I guess we'd need to do something like make a placeholder configmap with an empty string or something to get around that.

File: run.py (hunk 1 above)
Comment (NickGrantX) [reply run.py:33]: I don't know how to do that. Might have to pair on it to take advantage of your Pyhthonista skillz

**Review by LeoKHoward — APPROVED** [review]
Works for me

**Comment by NickGrantX** [comment]
> The testing instructions say to insert the IAP_AUDIENCE into the config for ops-ui, I assume you mean editing the deployed YAML? That works for testing it now but in CI for example it'll get overriden every time it's redeployed, it needs to be stored/fetched from a configmap. That then gives us the issue that it won't start up without it, I guess we'd need to do something like make a placeholder configmap with an empty string or something to get around that.

Comment addressed now

**Review by AdamHawtin — APPROVED** [review]
👍 Good proof of concept for when we have something here that needs audit logging.

Once the app actually has some features I'd be keen to move this into a middleware that audit logs every request.

**Pull request merged**